File: TNSAgent/tns/meter_point.py
# -*- coding: utf-8 -*- {{{
# vim: set fenc=utf-8 ft=python sw=4 ts=4 sts=4 et:

# Copyright (c) 2017, Battelle Memorial Institute
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# 'AS IS' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# The views and conclusions contained in the software and documentation
# are those of the authors and should not be interpreted as representing
# official policies, either expressed or implied, of the FreeBSD
# Project.
#
# This material was prepared as an account of work sponsored by an
# agency of the United States Government.  Neither the United States
# Government nor the United States Department of Energy, nor Battelle,
# nor any of their employees, nor any jurisdiction or organization that
# has cooperated in the development of these materials, makes any
# warranty, express or implied, or assumes any legal liability or
# responsibility for the accuracy, completeness, or usefulness or any
# information, apparatus, product, software, or process disclosed, or
# represents that its use would not infringe privately owned rights.
#
# Reference herein to any specific commercial product, process, or
# service by trade name, trademark, manufacturer, or otherwise does not
# necessarily constitute or imply its endorsement, recommendation, or
# favoring by the United States Government or any agency thereof, or
# Battelle Memorial Institute. The views and opinions of authors
# expressed herein do not necessarily state or reflect those of the
# United States Government or any agency thereof.
#
# PACIFIC NORTHWEST NATIONAL LABORATORY
# operated by BATTELLE for the UNITED STATES DEPARTMENT OF ENERGY
# under Contract DE-AC05-76RL01830

# }}}


# A MeterPoint may correlate directly with a meter. It necessarily
# corresponds to one measurement type (see MeasurementType enumeration) and
# measurement location within the circuit. Therefore, a single physical
# meter might be the source of more than one MeterPoint.

from datetime import datetime, date, timedelta

from measurement_type import MeasurementType
from measurement_unit import MeasurementUnit
from helpers import format_ts, format_date


class MeterPoint:
    # Text name that identifies a data file, database, or historian
    # where meter data points are to be stored. A default is provided,
    # but a meaningful filename should be used.
    today = date.today()
    dateFile = 'MP' + format_date(today)

    # DESCRIPTION
    description = ''

    # The measurement point datum that was collected during the last
    # reading update. This measurement must be of the measurement type
    # and measurement unit specified by object properties. The
    # measurement took place at the last update datetime. If needed
    # these measurements should be saved to a database or historian.
    lastMeasurement = None  # (1, 1)

    # Datetime of the last meter reading. This time is used with
    # the measurement interval to determine when the meter should be
    # read.
    lastUpdate = datetime.utcnow()

    # Constant time interval between readings. This and the last
    # measurement time are used to schedule the next meter reading.
    measurementInterval = timedelta(hours=1)

    # See MeasurementType enumeration. Property being metered.
    measurementType = MeasurementType.Unknown

    # See MeasurementUnit enumeration. Allowed units of measure. This
    # formulation is currently simplified by allowing only a limited
    # number of measurement units. These are not necessarily the raw
    # units; it should be the proper converted units.
    measurementUnit = MeasurementUnit.Unknown

    # NAME
    name = ''

    def __init__(self):
        pass

    def read_meter(self, obj):
        # Read the meter point at scheduled intervals
        #
        # MeterPoints are updated on a schedule. Properties have been defined to
        # keep track of the time of the last update and the interval between
        # updates.
        #
        # While this seems easy, meters will be found to be diverse and may use
        # diverse standards and protocols. Create subclasses and redefine this
        # function as needed to handle unique conditions.
        print('Made it to MeterPoint.read_meter() for ' + obj.name)

    def store(self):
        """
        Store last measurement into historian
        The default approach here could be to append a text record file. If the
        file is reserved for one meterpoint, little or no metadata need be
        repeated in records. Minimum content should be reading time and datum.
        Implementers will be found to have diverse practices for historians.
        """
        print('Made it to store() function in ' + self.name + ' targeting database ' + self.dataFile)

        # Open a simple text file for appending
        # Append the formatted, paired last measurement time and its datum
        with open(self.dataFile, "a") as myfile:
            myfile.write("{},{};".format(format_ts(self.lastUpdate), self.lastMeasurement))

File: Matlab/MeterPoint.m
classdef MeterPoint < handle
% MeterPoint Base Class
% A MeterPoint may correlate directly with a meter. It necessarily
% corresponds to one measurement type (see MeasurementType enumeration) and
% measurement location within the circuit. Therefore, a single physical
% meter might be the source of more than one MeterPoint.
     
%% MeterPoint Properties
    properties
        description = ''
        currentMeasurement = 0.0          % storage for last measured value
        currentMeasurementTime = datetime.empty
        lastStoreTime = datetime.empty
        % NOTE: Function Hours() corrects behavior of Matlab hours().
        measurementInterval = Hours(1)                               % [hr]
        measurementType = MeasurementType.unknown            %  enumeration
        measurementUnit = MeasurementUnit.unknown            %  enumeration
        name = ''
        nextStoreTime = datetime.empty
        status
        storeInterval = Hours(1)        
        writeFile = 'MP'+string(datetime('now','Format','yyMMdd'))
    end                                             % MeterPoint Properties
    
%% MeterPoint methods    
    methods

%% FUNCTION READ_METER()
function read_meter(obj)
% FUNCTION READ_METER() - Read a meter point at scheduled intervals
%
% This method will likely differ by application and meter type. 
% Pseudocode:
% 1. This method should be called when the measurementInterval has passed
%    since the currentMeasurementTime.
% 2. Read the meter (practices, protocols, and standards may differ).
% 3. Calculate and store the currentMeasurement.
% 4. Update currentMeasurementTime by adding the measurementInterval to the
%    currentMeasurementTime.
% 5. Update the nextMeasurementTime by adding the measurementInterval to
%    the nextMeasurmentTime.
%

    disp(['Made it to MeterPoint.read_meter() for ', obj.name]);
    
end                                                 % FUNCTION READ_METER()

%% FUNCTION STORE()
function store( obj )
% FUNCTION STORE() - Store the last measurement into a file
%   obj - MeterPoint object
% Pseudocode:
% 1. This method should be called when we are time storeInterval past the
%    lastStoreTime.
% 2. The currentMeasurement and its corresponding currentMeasurementTime
%    are appended to a file.
% 3. The lastStoreTime is updated by adding store interval.
    
%   Open a simple text file for appending    
    fid = fopen(obj.writeFile,'w');
    
%   Append the formatted, paired last measurement time and its datum
    fprintf(fid, ...
        string(datetime(obj.currentMeasurementTime,'Format', ...
        'yyMMdd:hhmmss'))+','+string(obj.currentMeasurement)+'\n');
    
%   Re-close the file
    fclose(fid);
    
%   Update the last storage time
%   NOTE: Function Hours() corrects the behavior of Matlab hours().
    obj.lastStoreTime = obj.lastStoreTime + Hours(obj.storeInterval);
    
%   Update the next storage time
    obj.nextStoreTime = obj.nextStoreTime + Hours(obj.storeInterval);

end                                                       %function store()
        
    end                                                % MeterPoint Methods
    
% Static MeterPoint Methods
    methods (Static)
        
%% TEST_ALL()                                                      COMPLETE
function test_all()
% TEST_ALL() - test all MeterPoint methods
    disp('Running MeterPoint.test_all()');
    MeterPoint.test_read_meter();
    MeterPoint.test_store();
end                                                              % TEST_ALL

%% TEST_READ_METER()
function pf = test_read_meter()
    disp('Running MeterPoint.test_read_meter()');
    pf = 'the test is not completed yet';
    disp(['WARNING: This test will be unique to the meter being read ',...
        'or the value being computed.']);

%   Success
    disp('- the test ran to completion');
    fprintf('Result: %s\n\n',pf);
    
% Clean up
%     clear test_mtr

end                                                     % TEST_READ_METER()

%% TEST_STORE()                                                    COMPLETE
function pf = test_store()
    disp('Running MeterPoint.test_store()');
    pf = 'pass';
    
    filename = 'MeterFileTest.txt';
    
%   Create a test meter and its writeFile
    test_mtr = MeterPoint;
        test_mtr.writeFile = filename;
        test_mtr.currentMeasurement = 3.1416;
        test_mtr.currentMeasurementTime = datetime(2018,1,1,12,0,0);
        test_mtr.lastStoreTime = datetime(2018,7,4,0,0,0);
%       NOTE: Function Hours() corrects behavior of Matlab hours().
        test_mtr.storeInterval = Hours(1);
        
    try
        test_mtr.store();
        disp('- the method ran without errors');
    catch
        pf = 'fail';
        error('- the method encountered errors and would not run');
    end
    
    if exist(filename,'file') ~= 2
        pf = 'fail';
        error('- the storage file does not exist');
    else
        disp('- the storage file was found to exist');
    end
    
    fid = fopen(filename);
    data = textread(filename, '%s', 'whitespace',',');
    fclose(fid);
    
    if str2num(data{2}) ~= 3.1416
        pf = 'fail';
        warning('- did not retreive the expected value');
    else
        disp('- retrieved the expected value');
    end

%   Success
    disp('- the test ran to completion');
    fprintf('Result: %s\n\n',pf);
    
% Clean up
    clear test_mtr
    fclose all;
%     oldstate = recycle('on');
%     delete 'MeterFileTest.txt'; %NOTE: This is not working for some reason.
%     recycle(oldstate);

end                                                          % TEST_STORE()


    end                                         % Static MeterPoint Methods
    
end                                                    %classdef MeterPoint

